src/bootstrap/llvm.rs

```diff
@@ -513,6 +513,45 @@ impl Step for Llvm {
     }
 }
 
+/// This step compiles LLVM and links it to an already compiled compiler sysroot.
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct LlvmLink {
+    pub target: TargetSelection,
+    pub compiler: Compiler,
+}
+
+impl Step for LlvmLink {
+    type Output = ();
+
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.alias("llvm-link")
```

Comment:
1) I wonder if someone might want to just build `llvm` using bootstrap, without having `rustc` built. `LlvmLink` now assumes that `rustc` has already been built - if it hasn't, it will probably error out (or be a no-op). With this change, it wouldn't be possible to successfully just build llvm.
2) `LlvmLink` is a bit special currently, it does not use the `Sysroot` step, which specially handles stage0 and downloaded rustc sysroot paths, but instead it hardcodes the sysroot path to `stageN`. I had to do this because `Sysroot` clears the `stageN` directory, and that must not happen if we want to avoid rebuilding `rustc`.

The logic in question is this one:
```rust
let sysroot_dir = |stage| {
    if stage == 0 {
        host_dir.join("stage0-sysroot")
    } else if builder.download_rustc() && compiler.stage != builder.top_stage {
        host_dir.join("ci-rustc-sysroot")
    } else {
        host_dir.join(format!("stage{}", stage))
    }
};
```
I could duplicate it in `LlvmLink` or put it into another separate step and use that one both in `Sysroot` and `LlvmLink`.

Reply:
> LlvmLink is a bit special currently, it does not use the Sysroot step, which specially handles stage0 and downloaded rustc sysroot paths, but instead it hardcodes the sysroot path to stageN. I had to do this because Sysroot clears the stageN directory, and that must not happen if we want to avoid rebuilding rustc.

oh, I see the problem - you want to reuse the sysroot *across invocations* of bootstrap, including when the config.toml has changed.

I wonder if you could do this today with something like `--keep-stage 1`, which avoids deleting the sysroot. I'm not sure if that currently copies over LLVM, but we could change it to do so, and that seems simpler than adding an "llvm link" step that actually means "reuse a previous sysroot" - I worry that the latter is going to be confusing and it won't be clear that it's actually *bypassing* the cache invalidation.

Reply:
When I have a locally built LLVM and then I simulate its "rebuild" with this:
```rust
$ rm build/x86_64-unknown-linux-gnu/llvm/llvm-finished-building && python3 x.py build library/std --keep-stage 1
```
then the compiler gets rebuilt, starting from `rustc_llvm`. The same situation is also a problem for the final dist build, where I would like to avoid the final rebuild of `rustc`. I'm not really sure what causes it, but when the LLVM artifacts change, the compiler is automatically recompiled. I suspected the build script of `rustc_llvm`, but it looks like it only watches the `llvm-wrapper` directory.